### ISSUE TITLE
MS SQL fail upset with primary key + composite index

### ISSUE THREAD

**Issue by joaoe** [sequelize/sequelize#13886]
## Issue Creation Checklist

[ ] I have read the [contribution guidelines](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)

### SSCCE

**Here is the link to the SSCCE for this issue:** LINK-HERE 

```js
let ItemModel = sequelize.define('Item', {
    id: {
      autoIncrement: true,
      type: DataTypes.INTEGER,
      allowNull: false,
      primaryKey: true,
      field: 'Id'
    },
    number: {
      type: DataTypes.INTEGER,
      allowNull: false,
      field: 'Code'
    },
    name: {
      type: DataTypes.STRING(200),
      allowNull: false,
      field: 'Name'
    },
    area: {
      type: DataTypes.INTEGER,
      allowNull: false,
      field: 'Area'
    }
  }, {
    sequelize,
    tableName: 'Item',
    schema: 'dbo',
    timestamps: false,
    indexes: [
      {
        name: 'IX_Item_U',
        unique: true,
        fields: [ 'Code', 'Name', 'Area' ]
      },
      {
        name: 'PK_Item',
        unique: true,
        fields: [ { name: 'Id' } ]
      },
    ]
  });

await sequelize.sync( {force:true})
let instance = await ItemModel.upsert({code: 101, name: 'me', area: 56});
```

### What do you expect to happen?

No, fail; upsert to run as expected.

### What is actually happening?

Fails with
```
Error: Primary Key or Unique key should be passed to upsert query
      at MSSQLQueryGenerator.upsertQuery (lib\dialects\mssql\query-generator.js:416:13)
      at MSSqlQueryInterface.upsert (lib\dialects\mssql\query-interface.js:69:37)
      at Function.upsert (lib\model.js:1494:46)
```

### Additional context

There are some small bugs in the upsert functions for the MS Sql dialect.

### Environment

- Sequelize version: main branch
- Node.js version:  16
- If TypeScript related: No

### How does this problem relate to dialects?

- [ ] I think this problem happens regardless of the dialect.
- [x] I think this problem happens only for the following dialect(s): mssql
- [ ] I don't know, I was using PUT-YOUR-DIALECT-HERE, with connector library version XXX and database version XXX

### Would you be willing to resolve this issue by submitting a Pull Request?

- [X] Yes, I have the time and I know how to start.
- [ ] Yes, I have the time but I don't know how to start, I would need guidance.
- [ ] No, I don't have the time, although I believe I could do it if I had the time...
- [ ] No, I don't have the time and I wouldn't even know how to start.

**Comment by ephys** [comment]
This works as expected. `upsert` needs a unique or PK attributes to be able to find the row that should be updated if it already exists.

If you're not providing either, just use `create` instead of `upsert`

**Comment by ephys** [comment]
Going to unassign myself, I initially misread part of the code sorry 😅.

However you're right. Other dialects seem to handle this case properly but mssql does not.

Can you work on a PR for this? :)

**Comment by joaoe** [comment]
> Can you work on a PR for this? :)

I have the fix ready.

**Comment by joaoe** [comment]
Just a heads up. During the week I have work and family to tend to, so I leave my github contributions for the weekends. I~m working on another PR, then I get to this one.

**Comment by rbWies** [comment]
I'm currently experiencing this as well; thought I had something in the options object wrong, but it apprears to be a bug. I appreciate an upcoming fix!

**Comment by joaoe** [comment]
@ephys Most dialects use the INSERT ON CONLIFCT statement to implement upsert, using the common code in abstract/query-generator.py.
But DB2 and MSSQL use a MERGE statement and each of those dialects have their own code to generate the MERGE statement even though the code is very similar and could be shred. The DB2 code also has the same bug.
I think the best is for these dialect to share the same code to generate the MERGE statement.

My question for you is here to best place it. I thought a bit about creating a `class QueryInterfaceWithMerge extends QueryInterface` which picks a new method `QueryGenerator.mergeQuery()` and that method does a feature check and has all common code. The method in QueryInterfaceWithMerge is the code currently in the mssql and db2 equivalent classes, with bugs fixed.

Any thoughts ? 

@rbWies I have a branch here ready in case you are in a hurry
https://github.com/joaoe/sequelize/tree/mssql_fix_upsert

**Comment by ephys** [comment]
If it's not too complicated I'd instead move the common code to a couple of utility functions that both dialect would call

**Comment by rbWies** [comment]
> @rbWies I have a branch here ready in case you are in a hurry but it will most likely change once it is merged https://github.com/joaoe/sequelize/tree/mssql_fix_upsert

Thank you!

I'm trying to install the specific fork and getting an error 

"npm ERR! package.json Non-registry package missing package.json: https://github.com/joaoe/sequelize/tree/mssql_fix_upsert.
npm ERR! package.json npm can't find a package.json file in your current directory."

I have a package.json in my directory... haven't been able to figure this out through research yet. I know it's not directly related, but any ideas?

**Comment by rbWies** [comment]
I manged to get it functioning; it looks like it was some build weirdness surrounding node-gyp and the fact that python was not installed on my system...

Upsert works! It's creating a PK "id" starting at 1000000000 and incrmenting from there, instead of taking the last highest value already in the column. Is that behaviour you coded in, or should I look elsewhere?

This lets me move on with development, so thank you very much for the patch. Do you know when it will be merged into the main release?

**Comment by joaoe** [comment]
>  It's creating a PK "id" starting at 1000000000 and incrmenting from there

That's unrelated with sequelize I think. Either you are passing that value somewhere, or the identity of that clumn is being initialized to 0.

To reset the identity value of a column to the lowest value possible (1 if the table is empty), run this

```
    DECLARE @table_name SYSNAME = 'MyTable';
    IF (SELECT last_value FROM sys.identity_columns c WHERE OBJECT_ID(@table_name) = c.object_id) IS NOT NULL
    BEGIN
        -- Reset to lowest value
        DBCC CHECKIDENT (@table_name, RESEED, 0);
        -- This call snaps the identity value to the next available
        DBCC CHECKIDENT (@table_name, RESEED);
    END
```

**Comment by joaoe** [comment]
> Do you know when it will be merged into the main release?

I have family, job and life in the way. I can spend some time on such projects during weekends or holiday.

**Comment by joaoe** [comment]
> if it's not too complicated I'd instead move the common code to a couple of utility functions that both dialect would call

@ephys  Ok, then I have a plan. I'll do the task once and I'll do it properly.
